openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,861 @@
+                                     OW_LITTLE_ENDIAN);
+         tempheader->headerLength = tempheader->headerLength+8;
+         break;
+      default:
+         return;
+      }
+
+   //retrieve the KeyIndex
+   if (msg->l2_keyIdMode != 0){
+      temp8b = *((uint8_t*)(msg->payload)+tempheader->headerLength);
+      msg->l2_keyIndex = (temp8b);
+      tempheader->headerLength = tempheader->headerLength+1;
+   } else {
+     //key is derived implicitly
+     msg->l2_keyIndex = 1;
+   }
+
+}
+
+/**
+\brief Identification of the key used to protect the frame and unsecuring operations.
+*/
+owerror_t IEEE802154_security_incomingFrame(OpenQueueEntry_t*      msg){
+   
+   m_deviceDescriptor*        deviceDescriptor;
+   m_keyDescriptor*           keyDescriptor;
+   m_securityLevelDescriptor* securityLevelDescriptor;
+   uint8_t nonce[13];
+   uint8_t i;
+   uint8_t myASN[5];
+   owerror_t outStatus;
+   uint8_t* a;
+   uint8_t len_a;
+   uint8_t* c;
+   uint8_t len_c;
+
+   //f key descriptor lookup
+   keyDescriptor = IEEE802154_security_keyDescriptorLookup(msg->l2_keyIdMode,
+                                                          &msg->l2_keySource,
+                                                          msg->l2_keyIndex,
+                                                          &msg->l2_keySource,
+                                                          idmanager_getMyID(ADDR_PANID),
+                                                          msg->l2_frameType);
+
+   if (keyDescriptor==NULL){
+      //can't find the key
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)2);
+      return E_FAIL;
+   }
+
+   //g device descriptor lookup
+   deviceDescriptor = IEEE802154_security_deviceDescriptorLookup(&msg->l2_keySource,
+                                                                idmanager_getMyID(ADDR_PANID),
+                                                                keyDescriptor);
+
+   if (deviceDescriptor==NULL){
+      //can't find the device
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)3);
+      return E_FAIL;
+   }
+
+   //h Security Level lookup
+   securityLevelDescriptor = IEEE802154_security_securityLevelDescriptorLookup(msg->l2_frameType,
+                                                                              msg->commandFrameIdentifier);
+
+   if (securityLevelDescriptor == NULL){
+	      //can't find the frame type
+	      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+	                           (errorparameter_t)msg->l2_frameType,
+	                           (errorparameter_t)4);
+	      return E_FAIL;
+   }
+
+   //i+j+k
+   if (IEEE802154_security_incomingSecurityLevelChecking(securityLevelDescriptor,
```

Comment:
I'm not a big fan of this construction, as the core of the behavior is hidden in the `if` statement.

I recommend you create an `outcome` variable, us that to hold the rc of `IEEE802154_security_incomingSecurityLevelChecking()`, then to the `if` on the `outcome` variable.